source/common/router/config_utility.cc

```diff
@@ -108,28 +109,13 @@ std::string ConfigUtility::parseDirectResponseBody(const envoy::config::route::v
     return EMPTY_STRING;
   }
   const auto& body = route.direct_response().body();
-  const std::string& filename = body.filename();
-  if (!filename.empty()) {
-    if (!api.fileSystem().fileExists(filename)) {
-      throw EnvoyException(fmt::format("response body file {} does not exist", filename));
-    }
-    const ssize_t size = api.fileSystem().fileSize(filename);
-    if (size < 0) {
-      throw EnvoyException(absl::StrCat("cannot determine size of response body file ", filename));
-    }
-    if (static_cast<uint64_t>(size) > max_body_size_bytes) {
-      throw EnvoyException(fmt::format("response body file {} size is {} bytes; maximum is {}",
-                                       filename, size, max_body_size_bytes));
-    }
-    return api.fileSystem().fileReadToEnd(filename);
-  }
-  const std::string inline_body(body.inline_bytes().empty() ? body.inline_string()
-                                                            : body.inline_bytes());
-  if (inline_body.length() > max_body_size_bytes) {
+
+  const std::string string_body = Envoy::Config::DataSource::read(body, true, api);
```

Comment:
Wow, this is much cleaner. Awesome!

I'm slightly concerned, though, about the potential for Envoy to run out of memory with this change. With the old code, in the case of a file we check the size before we allocate the string which contains it. With this new code, we'll read entire file into memory. This seems a bit risky. Is there any way we could get similar behavior? Maybe we should change DataSource::read() to take a max_size argument? I think this would be a fairly straightforward change and would avoid the OOM potential here.

Reply:
yes, but does it mean that anywhere in Envoy using `DataSource::read()` has this risk? Is there a goog way to fix them all?

Reply:
cc @RyanTheOptimist

Reply:
> yes, but does it mean that anywhere in Envoy using DataSource::read() has this risk? Is there a goog way to fix them all?

Yes, this is potentially an issue, though presumably only for data sources which are file-backed (and hence where the input is potentially enormous). I'm not sure it needs to block this PR, but it does seem worth filing an issue about.

Reply:
yeah, i think so, i will submit a new issue about the risk, and it will not be fixed in this pr, thanks.

Reply:
Well, I think we need to fix this *new* issue that we added in the PR since it changes behavior in a risky way. We can fix the *existing* issues in a followup.

Reply:
ok, so maybe we should not change `DataSource::read()` to take a max_size argument here? since it will affect anywhere using it. And what about keeping the file process code here, and then add `DataSource::read()` to support `environment_variable` and `inline string`?

Reply:
If we don't want to change all the callers of DataSource::read() now, then we could add the new max_size parameter with a default value of 0 (or -1, or some such). That seems like a pretty clean solution.

Reply:
yeah, i think it is a greater solution, i will fix it.